src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -221,66 +187,52 @@ internal SequencePosition Seek(in SequencePosition start, in SequencePosition en
             {
                 Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
                 var startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
 
                 int currentLength = startSegment.Memory.Length - startIndex;
+                
+                // Position in start segment, defer to single segment seek
                 if (currentLength > count)
-                {
-                    // Position in start segment, defer to single segment seek
-                    notInRange = false;
-                }
-                else
-                {
-                    return SeekMultiSegment(startSegment, startIndex, endObject, endIndex, count - currentLength);
-                }
-            }
-            else
-            {
-                Debug.Assert(startObject == endObject);
+                    goto IsSingleSegment;
+
+                // End of segment. Move to start of next.
+                return SeekMultiSegment(startSegment.Next, startIndex, endObject, endIndex, count - currentLength);
             }
 
-            if (notInRange)
+            Debug.Assert(startObject == endObject);
+
+            if (endIndex - startIndex < count)
                 ThrowHelper.ThrowArgumentOutOfRangeException_CountOutOfRange();
 
             // Single segment Seek
+        IsSingleSegment:
             return new SequencePosition(startObject, startIndex + (int)count);
         }
 
         [MethodImpl(MethodImplOptions.NoInlining)]
         private static SequencePosition SeekMultiSegment(ReadOnlySequenceSegment<T> currentSegment, int startIndex, object endObject, int endPosition, long count)
         {
             Debug.Assert(currentSegment != null);
-            Debug.Assert(currentSegment.Next != null);
             Debug.Assert(count >= 0);
 
-            // End of segment. Move to start of next.
-            currentSegment = currentSegment.Next;
-
-            if (count > 0 || currentSegment.Memory.Length == 0)
+            while (currentSegment != null && currentSegment != endObject)
             {
-                while (currentSegment != null)
-                {
-                    bool isCurrentAtEnd = currentSegment == endObject;
-                    int memoryLength = isCurrentAtEnd ? endPosition : currentSegment.Memory.Length;
-
-                    if (memoryLength > count || (memoryLength == count && isCurrentAtEnd))
-                    {
-                        // Fully contained in this segment; or at end of segment but there isn't a next one to move to
-                        break;
-                    }
-
-                    // Move to next
-                    count -= memoryLength;
-                    currentSegment = currentSegment.Next;
-                }
+                int memoryLength = currentSegment.Memory.Length;
+
+                // Fully contained in this segment; or at end of segment but there isn't a next one to move to
```

Comment:
This comment is not correct anymore without `memoryLength == count && isCurrentAtEnd` this check/